### PR TITLE
Fix memory error with ndenumerate and flat iterators.

### LINKED ISSUE #2212: Problems with numpy array iteration
I am getting bad behavior dealing with numpy-array iteration. The following shows an error when invoking `ndenumerate`. I get the same behavior when calling `.flat` on an array.  

## ndenumerate
```python
import numpy as np
import numba

def f():
    x = np.arange(6).reshape(2, 3)
    for index, item in np.ndenumerate(x):
        return index, item

jitted_f = numba.jit(f)
```
```python
>>> print(numba.__version__)
0.29.0

>>> print(f())
((0, 0), 0)

>>> print(jitted_f())
((0, 0), -2387225703656530210)
```

## flat
```python
def f():
    x = np.arange(6).reshape(2, 3)
    for item in x.flat:
        return item

jitted_f = numba.jit(f)
```
```python
>>> print(numba.__version__)
0.29.0

>>> print(f())
0

>>> print(jitted_f())
-2387225703656530210
```

This occurs in Python 3.5 for `numba 0.29.0`, and in Python 2.7 in `numba 0.29.0+0.gead0dd3.dirty`

### PR DESCRIPTION
**Pull request by sklam**
Fixes #2212 

Cause: reference to input array is not held by the iterators.

### CODE REVIEW

File: numba/datamodel/models.py [hunk 1]
```diff
@@ -984,7 +984,7 @@ def __init__(self, dmm, fe_type, need_indices):
         array_type = fe_type.array_type
         dtype = array_type.dtype
         ndim = array_type.ndim
-        members = [('array', types.EphemeralPointer(array_type)),
```
Comment (sklam) [inline numba/datamodel/models.py]: The pointee of a EphemeralPointer is not tracked; thus, no incref for it

File: numba/targets/arrayobj.py [hunk 2]
```diff
@@ -2790,11 +2790,10 @@ def make_array_flatiter(context, builder, arrty, arr):
     flatitercls = make_array_flat_cls(types.NumpyFlatType(arrty))
     flatiter = flatitercls(context, builder)
 
-    arrayptr = cgutils.alloca_once_value(builder, arr)
```
Comment (sklam) [inline numba/targets/arrayobj.py]: No need separate stack allocation anymore.  It is allocated as part of the iterator structure.

**Comment by pitrou** [comment]
Have you tried the performance impact of this change? Using non-refcounted pointers in array iterators was a deliberately introduced bug, in order to avoid increfs and decrefs at each iteration.

**Comment by sklam** [comment]
For both ndenumerate and flat iterators, no performance difference is observed.

Also, the incref and decref happens before and after the loop, and not in the loop body.

**Comment by pitrou** [comment]
> Also, the incref and decref happens before and after the loop, and not in the loop body.

Does it? Then my fears were unfounded. Good :-)

**Pull request merged**